### PR TITLE
Implement UDP discovery to get IP and push/sub info from server-game using UDP broadcast

### PR DESCRIPTION
**Pull request by miludmann**
+ some name refactoring and minor code cleanunp
+ Put default template of configuration.xml instead of a real config
+ Proxy can now start before server-game, order does not matter anymore
+ Implement timeouts

### CODE REVIEW

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            return attemptsPerformed < maxAttemptsNumber;
+        }
+        return (!udpBeaconDecoder.hasReceivedCorrectData() &&
+                attemptsPerformed < maxAttemptsNumber);
+    }
+
+    /**
+     * Find the server using UDP broadcast and fill data fields
+     */
+    public void startBroadcasting() {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: This does more than start to broadcast. You may want to change the name and split into more methods.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                        continue; // Do not broadcast to the loopback interface
+                    }
+
+                    for (final InterfaceAddress interfaceAddress : networkInterface.getInterfaceAddresses()) {
+                        final InetAddress broadcast = interfaceAddress.getBroadcast();
+                        if (null == broadcast) {
+                            continue;
+                        }
+
+                        // Send the broadcast package
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: For instance, this command can nearly become the name of the method doing what comes next.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                            logback.error(e.getMessage());
+                        }
+
+                        logback.info("Request packet sent to: " + broadcast.getHostAddress() + "; Interface: " + networkInterface.getDisplayName());
+                    }
+                }
+
+                logback.info("Done looping over all network interfaces. Now waiting for a reply!");
+
+                // Wait for a response
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: Here again this code can be put in a method.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    private final int broadcastPort;
+    private final int receiverBufferSize = 512;
+    private final DatagramSocket datagramSocket;
+    private final UdpBeaconDecoder udpBeaconDecoder;
+    private int maxAttemptsNumber = 1;
+    private int attemptsPerformed = 0;
+    private String pushAddress;
+    private String subscribeAddress;
+
+    public UdpBeaconFinder(final DatagramSocket datagramSocket, final int broadcastPort, final UdpBeaconDecoder udpBeaconDecoder) {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: This line is a bit long.

File: lejos-mf-pc [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit 3e4282d467ef284ffc03c5d872fb238d07a3e608
+Subproject commit 97883cbacf0b9fe8aad866fd2241917b87cda6c0
```
Comment (dchilot) [inline lejos-mf-pc]: What did you gain from this upgrade ?

Comment (miludmann) [inline lejos-mf-pc]: It was indeed a mistake, well spotted! This should not be here, the version is not good... Will update it.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java (hunk 2 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: Probably clearer yes.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java (hunk 1 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: Right! Will be modified.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java (hunk 3 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: Will do.

File: proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java (hunk 4 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/UdpBeaconFinder.java]: Will add some CL LR

File: proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java [hunk 6]
```diff
@@ -218,7 +240,8 @@ public void run() {
             // We stop the service once there are no more robots
             // connected to the proxy
             while (!Thread.currentThread().isInterrupted() &&
-                    !robotsMap.getConnectedRobots().isEmpty()) {
+                    !robotsMap.getConnectedRobots().isEmpty() &&
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java]: This might deserve a dedicated method to clarify.

File: proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java [hunk 7]
```diff
@@ -235,8 +258,12 @@ public void run() {
                 }
             }
             logback.info("End of communication service");
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java]: Maybe a terminateCommunicationService method ?

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java [hunk 8]
```diff
@@ -9,4 +13,34 @@
...
+     */
+    public static List<byte[]> split(final byte separator, final byte[] input, final int limit) {
+        final List<byte[]> list = new LinkedList<>();
+        if (0 == limit) {
+            return list;
+        }
+
+        int blockStart = 0;
+        if (1 != limit) {
+            int position = 0;
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: Could this be that position is a kind of blockEnd ?

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java [hunk 9]
```diff
@@ -9,4 +13,34 @@
...
+            final boolean limited = 0 < limit;
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: Reading !limited might not be as easy as reading unlimited.

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java [hunk 10]
```diff
@@ -9,4 +13,34 @@
...
+     * @return list of bytes arrays, it size being of maximum {limit} elements
+     */
+    public static List<byte[]> split(final byte separator, final byte[] input, final int limit) {
+        final List<byte[]> list = new LinkedList<>();
+        if (0 == limit) {
+            return list;
+        }
+
+        int blockStart = 0;
+        if (1 != limit) {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: If the limit is set to 1 you split the list in one. It should return a copy of the input. If it is set to 0, you get the exact input. There is something a bit weird.

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java [hunk 11]
```diff
@@ -9,4 +13,34 @@
...
     }
 
+    /**
+     * Split a byte array around a byte separator, within a given limit
+     *
+     * @param separator byte separator
+     * @param input     bytes array to split
+     * @param limit     max number of elements allowed in the returned list
+     *                  If limit is < 0, then proceed as if there were no limit
+     * @return list of bytes arrays, it size being of maximum {limit} elements
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: Typo: "it size" -> "its size"

File: proxy-robots-module/src/main/java/orwell/proxy/config/Configuration.java [hunk 12]
```diff
@@ -35,19 +37,13 @@ private void populate() {
 
             switch (enumConfigFileType) {
                 case RESOURCE:
-                    final InputStream xml = getClass().getResourceAsStream(filePath);
-                    this.configuration = (ConfigModel) unmarshaller.unmarshal(xml);
-                    isPopulated = true;
-                    logback.info("Configuration loaded from resource file");
+                    onResource(unmarshaller);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/config/Configuration.java]: Maybe populateFromResource ?

File: proxy-robots-module/src/main/java/orwell/proxy/config/Configuration.java [hunk 13]
```diff
@@ -35,19 +37,13 @@ private void populate() {
...
                     break;
                 case FILE:
-                    final File file = new File(filePath);
-                    this.configuration = (ConfigModel) unmarshaller.unmarshal(file);
-                    isPopulated = true;
-                    logback.info("Configuration loaded from external file");
+                    onFile(unmarshaller);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/config/Configuration.java]: Maybe populateFromFile ?

File: proxy-robots-module/src/main/java/orwell/proxy/config/Configuration.java [hunk 14]
```diff
@@ -35,19 +37,13 @@ private void populate() {
...
                     break;
                 default:
-                    logback.error("Config file type of " + enumConfigFileType + " not handled");
+                    onDefault();
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/config/Configuration.java]: And now you might think that I would suggest something like populateFromNotHandled, but no. I would suggest to create a dedicated object for file, resource, url (and whatnot) thus invalidating the previous two comments. This would allow you to get rid of enumConfigFileType, that is so hard to track and use polymorphism instead.

File: proxy-robots-module/src/main/java/orwell/proxy/robot/IPWebcam.java [hunk 15]
```diff
@@ -13,12 +13,12 @@
     private final URL url;
 
     public IPWebcam(final URL url) {
-        assert(null != url);
+        assert (null != url);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/IPWebcam.java]: I am quite used to consider assert a normal function. Why did you add a space ? Is it a Java style thing ?

File: proxy-robots-module/src/main/java/orwell/proxy/robot/RobotGameStateVisitor.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    private void setGameState(final IRobot robot) {
+        switch (gameState) {
+            case WAITING_TO_START:
+                setWaitingToStart(robot);
+                break;
+            case PLAYING:
+                setPlaying(robot);
+                break;
+            case FINISHED:
+                if (null == winningTeam || winningTeam.equals("")) {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/RobotGameStateVisitor.java]: This might fit well in a dedicated method.

File: proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+                    final NetworkInterface networkInterface = interfaces.nextElement();
+
+                    if (networkInterface.isLoopback() || !networkInterface.isUp()) {
+                        continue; // Do not broadcast to the loopback interface
+                    }
+
+                    for (final InterfaceAddress interfaceAddress : networkInterface.getInterfaceAddresses()) {
+                        final InetAddress broadcastAddress = interfaceAddress.getBroadcast();
+                        if (null != broadcastAddress) {
+                            logback.info("Trying to send broadcast package on interface: " + networkInterface.getDisplayName());
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java]: I think there was a comment earlier about how far this goes but It really is indented a lot.

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBOM.java [hunk 18]
```diff
@@ -29,42 +31,47 @@ public ZmqMessageBOM(final String routingId, final EnumMessageType messageType,
...
-     *                        routingID typeString message
+     *                        routingId typeString message
+     *                        Careful: message is arbitrary binary data
      */
     public static ZmqMessageBOM parseFrom(final byte[] raw_zmq_message) throws ParseException {
-        final String zmqMessageString = new String(raw_zmq_message);
-        final String[] zmqMessageStringArray = zmqMessageString.split(" ", 3);
+        // We do not want to create a String from arbitrary binary data, so we
+        // first isolate the 3 parts of the raw zmq message
+        final List<byte[]> zmqMessageBytesList = Utils.split(ZMQ_SEPARATOR, raw_zmq_message, 3);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBOM.java]: There is a magic number and it explanation which suggests a dedicated constant would be better.

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java [hunk 19]
```diff
@@ -160,11 +170,35 @@ public void run() {
                     logback.error("ZmqReader thread sleep exception: " + e.getMessage());
                 }
             }
+            if (hasSocketRecvTimeout()) {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java]: Maybe the whole end this method could be in a separate method.

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java [hunk 20]
```diff
@@ -131,25 +142,24 @@ public int getNbMessagesSkipped() {
         @Override
         public void run() {
             logback.info("ZmqReader has been started");
-            while (isConnected) {
+            baseTimeMs = System.currentTimeMillis();
+            while (isConnected && !hasSocketRecvTimeout()) {
                 final byte[] raw_zmq_message = receiver.recv(ZMQ.NOBLOCK);
                 if (null != raw_zmq_message) {
                     synchronized (rXguard) {
+                        baseTimeMs = System.currentTimeMillis();
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java]: Maybe the synchronised inner block could be in a dedicated method.

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java [hunk 21]
```diff
@@ -120,8 +122,17 @@ public boolean isConnectedToServer() {
...
-        return nbMessagesSkipped;
+    public int getNbSuccessiveMessagesSkipped() {
+        return nbSuccessiveSameMessages;
+    }
+
+    public int getNbBadMessages() {
+        return nbBadMessages;
+    }
+
+    private boolean hasSocketRecvTimeout() {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java]: I am a bit uncomfortable with the name of the method. Maybe hasReceiveTimeoutExpired ?

File: proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java [hunk 22]
```diff
@@ -0,0 +1,335 @@
...
+        replay(mockedDatagramSocket);
+
+        udpBeaconFinder = new UdpBeaconFinder(mockedDatagramSocket, BROADCAST_TIMEOUT_MS, udpBeaconDecoder);
+        udpBeaconFinder.broadcastAndGetServerAddress();
+        logback.info(udpBeaconFinder.toString());
+
+        verify(mockedDatagramSocket);
+        assertTrue(udpBeaconDecoder.hasReceivedCorrectData());
+        assertEquals(IP_TEST, udpBeaconDecoder.getServerGameIp());
+        assertEquals("tcp://127.0.0.1:9001", udpBeaconDecoder.getPushAddress());
```
Comment (dchilot) [inline proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java]: Maybe those two deserve dedicated constants ?

File: proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+import static org.junit.Assert.*;
+
+/**
+ * Created by Michaël Ludmann on 5/26/15.
+ */
+@SuppressWarnings("unused")
+@RunWith(JUnit4.class)
+public class UdpBeaconTest {
+    private final static Logger logback = LoggerFactory.getLogger(UdpBeaconTest.class);
+    private final static int BROADCAST_PORT = 9080;
```
Comment (dchilot) [inline proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java]: It looks like it is not used.

File: proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+                                       @Override
+                                       public Object answer() throws Throwable {
+                                           final DatagramPacket datagramPacket = (DatagramPacket) getCurrentArguments()[0];
+                                           datagramPacket.setAddress(InetAddress.getByName(IP_TEST));
+                                           datagramPacket.setData(getTestPacketDataBytes(BAD_BYTE, ADDRESS_SIZE, PARTIAL_PUSH_PORT, CHECK_BYTE_SEPARATOR, ADDRESS_SIZE, PARTIAL_PUB_PORT, CHECK_BYTE_ETX));
+                                           return null;
+                                       }
+                                   }
+                // Then return a good one at the second try
+        ).andAnswer(new IAnswer<Object>() {
```
Comment (dchilot) [inline proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java]: The difference in indentation for the two answers is a bit unsettling.

File: proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java (hunk 6 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java]: Will do.
As for the sleep inside the thread, I am taking advice from https://docs.oracle.com/javase/tutorial/essential/concurrency/sleep.html

In practice, if I do not put such a sleep, then I make one of my cores 100% busy just for this one thread...

File: proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java (hunk 7 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/ProxyRobots.java]: Alrighty! :)

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java (hunk 8 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: Clearer yes!

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java (hunk 9 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: I am not a big fan of negations in names... I try to never to it for consistency and clarity.

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java (hunk 10 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: Actually, no.

if limit = 0: returns a new empty list
if limit = 1: returns the first element result of the split (more precisely: a list of one byte array of all the bytes encountered in the input until the first separator)

I am adding this comment in the javadoc

File: proxy-robots-module/src/main/java/orwell/proxy/Utils.java (hunk 11 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/Utils.java]: Thanks!

File: proxy-robots-module/src/main/java/orwell/proxy/robot/IPWebcam.java (hunk 15 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/IPWebcam.java]: assert is a language feature, not a function. But in a way, you are right indeed, parenthesis are confusing and useless (a bit like adding parenthesis after a return () ).
I will remove parenthesis in such assertions.

File: proxy-robots-module/src/main/java/orwell/proxy/robot/RobotGameStateVisitor.java (hunk 16 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/RobotGameStateVisitor.java]: +1

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBOM.java (hunk 18 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBOM.java]: Will add the constant instead!

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java (hunk 19 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java]: Will put this in a new method: terminateZmqReader()

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java (hunk 20 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java]: I will put this block in a method: handleRawZmqMessage(raw_zmq_message);

File: proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java (hunk 21 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/zmq/ZmqMessageBroker.java]: I almost agree; so I will rename it to hasReceivedTimeoutExpired ;)

File: proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java (hunk 22 above)
Comment (miludmann) [inline proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java]: Da

File: proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java (hunk 23 above)
Comment (miludmann) [inline proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java]: Well spotted!

File: proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java (hunk 24 above)
Comment (miludmann) [inline proxy-robots-module/src/test/java/orwell/proxy/udp/UdpBeaconTest.java]: I'll make this uniform

File: proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java (hunk 17 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java]: I will then add some methods.

**Comment by massix** [comment]
It might be just me, but I have a feeling this PR goes far way beyond its original purpose.
Could you please consider splitting it in multiple PR having different feature-oriented branches, so that the integrators won't have to deal with three thousand features all at once?

Please do not hesitate to contact us again if the problem persists.

Best regards.

**Comment by miludmann** [comment]
Yep, I reckon I might have been misusing Github lately: I indeed do all my dev on feature branches (check my fork), yet once done I was then merging them on my master, which inadvertently updates the PR as well, without my willing consent...
I'll try to find a way to do it in cleaner way later on.

Anyway, the integrator has been following this PR quite closely, so I trust him in managing this monster PR well :)

**Comment by miludmann** [comment]
This should be it. All the integrator comments have been taken into account.

This PR will mark the end of Milestone 1 \o/

File: proxy-robots-module/src/main/java/orwell/proxy/config/source/ConfigurationFile.java [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+public class ConfigurationFile extends Configuration {
+    private final static Logger logback = LoggerFactory.getLogger(ConfigurationFile.class);
+    private final String filePath;
+
+    public ConfigurationFile(final String filePath) throws FileNotFoundException {
+        this.filePath = filePath;
+
+        final File file = new File(filePath);
+        if (!file.exists() || file.isDirectory()) {
+            throw new FileNotFoundException(filePath);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/config/source/ConfigurationFile.java]: If it is a directory, it is not really missing (it will not make a big difference).

Comment (massix) [inline proxy-robots-module/src/main/java/orwell/proxy/config/source/ConfigurationFile.java]: We should probably throw a more meaningful exception and we could have a factory for the similar ones.

Something like
```java
throw new FileOrDirectoryExceptionsFactory.getFactory(File file).throwAdequateExceptionPlease()
```

File: proxy-robots-module/src/main/java/orwell/proxy/robot/RfidSensor.java [hunk 26]
```diff
@@ -93,4 +89,18 @@ public void clear() {
     public void accept(final IRobotElementVisitor visitor) {
         visitor.visit(this);
     }
+
+    @Override
+    public String toString() {
+        final StringBuilder stringBuilder = new StringBuilder("RfidSensor { ");
+        for (final Robot.Rfid rfid : this.rfidSensorReads) {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/RfidSensor.java]: Could you please confirm if this Robot is coming from orwell.messages.Robot ?

File: proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+    public int getNumberOfPerformedAttempts() {
+        return attemptsPerformed;
+    }
+
+    /**
+     * @return true if we did not get any meaningful response from server
+     * and have not reached max allowed attempts number
+     */
+    private boolean shouldTryToFindBeacon() {
+        if (null == udpBeaconDecoder) {
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java]: Does it mean anything to have udpBeaconDecoder null ? UdpBeaconFinder seems to be constructed with a non null object, and it seems that udpBeaconDecoder is needed to have the expected behaviour.

File: proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+        logback.info("Request packet sent to: " + broadcastAddress.getHostAddress());
+    }
+
+    private void waitForServerResponse(final DatagramSocket datagramSocket) throws IOException {
+        final byte[] recvBuf = new byte[receiverBufferSize];
+        final DatagramPacket receivePacket = new DatagramPacket(recvBuf, recvBuf.length);
+
+        try {
+            datagramSocket.receive(receivePacket);
+            udpBeaconDecoder.parseFrom(receivePacket);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java]: NullPointer exception in case udpBeaconDecoder is null ?

File: proxy-robots-module/src/main/java/orwell/proxy/robot/RfidSensor.java (hunk 26 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/RfidSensor.java]: I confirm it.

File: proxy-robots-module/src/main/java/orwell/proxy/config/source/ConfigurationFile.java (hunk 25 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/config/source/ConfigurationFile.java]: Ok, will add NotFileFoundException.
(Java already has a NotDirectoryFoundException)

File: proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java (hunk 27 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java]: You are right, I will add an assert != null in the constructor and remove the other checks.

File: proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java (hunk 28 above)
Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/udp/UdpBeaconFinder.java]: Not needed anymore (cf. previous comment)

File: proxy-robots-module/src/main/java/orwell/proxy/GameState.java [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Created by Michaël Ludmann on 6/4/15.
+ */
+public class GameState {
+    private final static Logger logback = LoggerFactory.getLogger(GameState.class);
+    private ServerGame.GameState serverGameGameState;
+
+    public GameState(final byte[] gameStateMessage) {
+        try {
+            this.serverGameGameState = ServerGame.GameState.parseFrom(gameStateMessage);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/GameState.java]: Can serverGameGameState be null ?
It seems that the visitor will fail if this is null.

Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/GameState.java]: Yes, it is allowed to be null (in case of a ParseError, which might happen). In this case, the visitor will be ignored by the class visited. Cf RobotsMap :+1:     @Override
    public void accept(final RobotGameStateVisitor visitor) {
        if (null != visitor)
            visitor.visit(this);
    }

File: proxy-robots-module/src/main/java/orwell/proxy/robot/ColourSensor.java [hunk 30]
```diff
@@ -85,4 +81,18 @@ public void clear() {
     public void accept(final IRobotElementVisitor visitor) {
         visitor.visit(this);
```
Comment (dchilot) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/ColourSensor.java]: Some visitors cannot be null while some others can, I find this quite confusing.

Comment (miludmann) [inline proxy-robots-module/src/main/java/orwell/proxy/robot/ColourSensor.java]: Ok then, I agree it is not good consistency wise.
I'll remove the null check for the visitor (which anyway won't fail since it will fall in the DEFAULT-UNDEFINED case which is handled)
